server/src/main/java/com/linecorp/centraldogma/server/internal/api/RepositoryServiceV1.java

```diff
@@ -93,29 +86,12 @@ public RepositoryServiceV1(ProjectManager projectManager, CommandExecutor execut
      * <p>Creates a new repository.
      */
     @Post("/projects/{projectName}/repos")
-    public CompletionStage<RepositoryDto> createRepository(@Param("projectName") String projectName,
-                                                           @RequestObject JsonNode node) {
-        checkProjectExists(projectName);
-
-        final String repoName = getName(node);
-        return execute(Command.createRepository(AuthenticationUtil.currentAuthor(), projectName, repoName))
-                .handle((unused, thrown) -> {
-                    if (thrown != null) {
-                        if (Throwables.getRootCause(thrown) instanceof StorageExistsException) {
-                            throw newHttpResponseException(HttpStatus.CONFLICT,
-                                                           "repository " + repoName + " already exists");
-                        }
-                        return Exceptions.throwUnsafely(thrown);
-                    }
-                    return DtoConverter.convert(getRepository(projectName, repoName));
-                });
-    }
-
-    private static String getName(JsonNode node) {
-        checkArgument(node.get("name") != null, "repository name should be non-null");
-        final String repoName = node.get("name").textValue();
-        checkArgument(!isNullOrEmpty(repoName), "repository name should be non-null");
-        return repoName;
+    public CompletionStage<RepositoryDto> createRepository(@RequestObject Project project,
+                                                           @RequestObject CreateRepositoryRequest request,
+                                                           @RequestObject Author author) {
+        return mds.addRepo(project.name(), author, request.name())
+                  .thenCompose(p -> execute(Command.createRepository(author, p.name(), request.name())))
+                  .handle(returnOrThrow(() -> DtoConverter.convert(project.repos().get(request.name()))));
```

Comment:
Don't we need to remove the metadata about repository in case it fails to create the repository?

Reply:
Oops, fixed. Thanks!